### ISSUE TITLE
Mask RCNN: No masks generated

### ISSUE THREAD

**Issue by andreap9** [matterport/Mask_RCNN#2122]
Hi,

I did a training with my own data  (136 images, 20 Epochs, 600 steps per Epoch) to detect particles in images, but when I run the detection I obtain too big and too many bounding boxes and no masks. What could be the problem? Could this be an underfitting or overfitting issue? How can I fix it? The results are very bad also for images that I already trained. I attach photo:
![visualize](https://user-images.githubusercontent.com/60215950/79474027-34aaf680-8006-11ea-8385-b9b0e45ad88f.png)

**Comment by ejri** [comment]
Hey, I'm having similar issues. Did you manage to figure out what the issues are?

**Comment by andreap9** [comment]
> Hey, I'm having similar issues. Did you manage to figure out what the issues are?

From the learning curves I got I guess is because I don´t have enough training data (I am using around 150 images), so it would be an underfitting problem. Now I am trying to generate more data and see if after it will work better.

**Comment by ejri** [comment]
I'm at about 700 images, though some people were having OK results with under 100 images.

I saw in other threads that it could also be to the scipy version, so I've tried 1.2.2, and I got this. 

![download](https://user-images.githubusercontent.com/2915475/79907356-502f4c00-83ce-11ea-842d-b1c43b4c4268.png)

There was another thread about some possible issues with the instances having integer values, when they are supposed to be floats. 

(https://github.com/matterport/Mask_RCNN/issues/912)

**Comment by andreap9** [comment]
> 1.2.2

I downgraded the scipy version but I still got the same results. However I modified 
utils.py line 866 from to shift = np.array([0, 0, 1., 1.]) and now I have less bounding boxes:
![visualize](https://user-images.githubusercontent.com/60215950/79973575-188cd680-8498-11ea-9a7c-6860d3bcd6af.png)

**Comment by truongtd6285** [comment]
I guess that the size of training images are much smaller than the ones you used for inference?

**Comment by andreap9** [comment]
> I guess that the size of training images are much smaller than the ones you used for inference?

No, they all have the same size and the picture I uploaded is one of the images I used for training

**Comment by suchiz** [comment]
Hi @andreap9, im not sure about modifying the code of Mask RCNN. If you want less detection, just change the detection threshold in the config. I can see that you allow objects with confidences of 0.6 which are really low if you dont have many classes to detect. The lower is the threshold, the more noise(false positive) you'll get.

**Comment by ejri** [comment]
<img width="177" alt="Screen Shot 2020-05-11 at 11 47 09 PM" src="https://user-images.githubusercontent.com/2915475/81647407-d0e7d380-93e1-11ea-9230-58f93758dbb9.png">

<img width="212" alt="Screen Shot 2020-05-11 at 11 47 17 PM" src="https://user-images.githubusercontent.com/2915475/81647403-cf1e1000-93e1-11ea-9c72-a8d46b769d13.png">



@andreap9 It took us a lot more training (a lot more images and more epochs) to get relatively ok results, because we were training on 13 classes. Start by training one class, play around with the threshold, see if that helps

**Comment by suchiz** [comment]
Hi @ejri Looks like your BB are goods ! Congrats, nonetheless you probably have a problem with the masks. Even thought you didn't train much. The result looks like there is a problem. Maybe check your input masks (are they well loaded?) and maybe again, as you said, the scipy version.

**Comment by andreap9** [comment]
> Hi @andreap9, im not sure about modifying the code of Mask RCNN. If you want less detection, just change the detection threshold in the config. I can see that you allow objects with confidences of 0.6 which are really low if you dont have many classes to detect. The lower is the threshold, the more noise(false positive) you'll get.

Thank you for your answer! I generated more training images and now it works.

**Comment by andreap9** [comment]
> <img alt="Screen Shot 2020-05-11 at 11 47 09 PM" width="177" src="https://user-images.githubusercontent.com/2915475/81647407-d0e7d380-93e1-11ea-9230-58f93758dbb9.png">
> 
> <img alt="Screen Shot 2020-05-11 at 11 47 17 PM" width="212" src="https://user-images.githubusercontent.com/2915475/81647403-cf1e1000-93e1-11ea-9c72-a8d46b769d13.png">
> 
> @andreap9 It took us a lot more training (a lot more images and more epochs) to get relatively ok results, because we were training on 13 classes. Start by training one class, play around with the threshold, see if that helps

Thank you for your answer! You were right about the threshold, but the problem was that the only detections I was getting were that ones with around 0.6 confidence. Anyway, my problem was an underfitting. I didnt have enough training images and my validation dataset was also too small. Now my dataset is much bigger and it works!

**Comment by kimile599** [comment]
> > <img alt="Screen Shot 2020-05-11 at 11 47 09 PM" width="177" src="https://user-images.githubusercontent.com/2915475/81647407-d0e7d380-93e1-11ea-9230-58f93758dbb9.png">
> > <img alt="Screen Shot 2020-05-11 at 11 47 17 PM" width="212" src="https://user-images.githubusercontent.com/2915475/81647403-cf1e1000-93e1-11ea-9c72-a8d46b769d13.png">
> > @andreap9 It took us a lot more training (a lot more images and more epochs) to get relatively ok results, because we were training on 13 classes. Start by training one class, play around with the threshold, see if that helps
> 
> Thank you for your answer! You were right about the threshold, but the problem was that the only detections I was getting were that ones with around 0.6 confidence. Anyway, my problem was an underfitting. I didnt have enough training images and my validation dataset was also too small. Now my dataset is much bigger and it works!

How large is your dataset right now and did you have onject have different scales in the picture. I mean in a the same resolution like 512*512, one object might cover 300 pixels while others might only cover 50 pixels. 

Thank you.

**Comment by kaypositive** [comment]
Please how did you solve this problem. I have similar challenge. labels, boxes and mask are totally off and sometimes at the corners of the image. Mask results are boolean I thought they should be floats numbers representing x,y coordinates

**Comment by TianRuoyu** [comment]
Did you solve this problem？

**Comment by Unmicool** [comment]
> Did you solve this problem？


> Please how did you solve this problem. I have similar challenge. labels, boxes and mask are totally off and sometimes at the corners of the image. Mask results are boolean I thought they should be floats numbers representing x,y coordinates

Maybe this comment can solve your problem: https://github.com/matterport/Mask_RCNN/issues/2822#issuecomment-1129205788
